### PR TITLE
trivy-sbom

### PR DESCRIPTION
**Pull request by anila-intelops**
trivy-sbom

### CODE REVIEW

File: model/trivy_sbom.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package model
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+)
+
+//import "github.com/aquasecurity/trivy/pkg/types"
```
Comment (avinashkna4) [inline model/trivy_sbom.go:9]: remove it

File: agent/kubviz/k8smetrics_agent.go [hunk 2]
```diff
@@ -146,14 +151,15 @@ func main() {
 				}
 			}
 		}()
-		wg.Add(7) // Initialize the WaitGroup for the seven goroutines
+		wg.Add(8) // Initialize the WaitGroup for the seven goroutines
```
Comment (avinashkna4) [inline agent/kubviz/k8smetrics_agent.go:154]: its 9 routines. wg.Add(9)

File: agent/kubviz/trivy_sbom.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	"fmt"
+	"log"
+	"os/exec"
+	"sync"
+	"time"
+
+	//"github.com/aquasecurity/trivy/pkg/k8s/report"
+	// "github.com/aquasecurity/trivy/pkg/types"
+	//"github.com/aquasecurity/trivy/pkg/fanal/types"
+	// "github.com/aquasecurity/trivy/pkg/sbom/cyclonedx/core"
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:15]: remove it

File: agent/kubviz/trivy_sbom.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+func RunTrivySbomScan(config *rest.Config,js nats.JetStreamContext, wg *sync.WaitGroup, errCh chan error) {
+	defer wg.Done()
+    images, err := ListImages(config)
+	log.Println("length of images", len(images))
+
+	if err != nil {
+		log.Fatalf("failed to list images: %v", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:59]: 300 sec may not be enough sometimes. Lets have it as 600sec.

File: agent/kubviz/trivy_sbom.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	return stdout, nil
+}
+
+func RunTrivySbomScan(config *rest.Config,js nats.JetStreamContext, wg *sync.WaitGroup, errCh chan error) {
+	defer wg.Done()
+    images, err := ListImages(config)
+	log.Println("length of images", len(images))
+
+	if err != nil {
+		log.Fatalf("failed to list images: %v", err)
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:56]: not Fatal, it will kill the application.

File: agent/kubviz/trivy_sbom.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		// Start a goroutine for each image
+		go func(i int, image model.RunningImage) {
+			defer wgc.Done()
+
+			// Execute the Trivy command with the context
+			command := fmt.Sprintf("trivy image --format cyclonedx %s", image.PullableImage)
+			out, err := executeCommandSbom(ctx, command)
+
+			if ctx.Err() == context.DeadlineExceeded {
+				log.Printf("*************Command execution timeout for image %s", image.PullableImage)
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:77]: no stars

File: agent/kubviz/trivy_sbom.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+			}
+
+			if err != nil {
+				log.Printf("*************Error executing Trivy for image %s: %v", image.PullableImage, err)
+				return // Move on to the next image in case of an error
+			}
+
+			// Check if the output is empty or invalid JSON
+			if len(out) == 0 {
+				log.Printf("*************Trivy output is empty for image %s", image.PullableImage)
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:88]: no stars

File: agent/kubviz/trivy_sbom.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	metrics := model.Reports{
+		ID:     uuid.New(),
+		Report: report,
+	}
+	metricsJson, _ := json.Marshal(metrics)
+	_, err := js.Publish(constants.TRIVY_SBOM_SUBJECT, metricsJson)
+	if err != nil {
+		errCh <- err
+	}
+	//log.Println("Report", report)
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:33]: remove it

File: agent/kubviz/trivy_sbom.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+func executeCommandSbom(ctx context.Context, command string) ([]byte, error) {
+	cmd := exec.CommandContext(ctx, "/bin/sh", "-c", command)
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		log.Println("Execute Command Error", err.Error())
+	}
+	// Print the output
+	// log.Println(string(stdout))
```
Comment (avinashkna4) [inline agent/kubviz/trivy_sbom.go:46]: remove it

File: client/pkg/clickhouse/statements.go [hunk 10]
```diff
@@ -145,6 +145,18 @@ const dockerHubBuildTable DBStatement = `
...
+	CREATE TABLE IF NOT EXISTS trivysbom (
+		id UUID,
+		schema String,
+		bom_format String,
+		spec_version String,
+		serial_number String,
+		version BIGINT,
+		metadata_timestamp DateTime('UTC'),
+		vulnerabilities Array(String)
+	) engine=File(TabSeparated)
```
Comment (avinashkna4) [inline client/pkg/clickhouse/statements.go:158]: this data is not enough i guess. We need components and metadata as well. and make each vulnerability as row.

File: model/trivy_sbom.go (hunk 1 above)
Comment (anila-intelops) [reply model/trivy_sbom.go]: removed comment line

File: agent/kubviz/k8smetrics_agent.go (hunk 2 above)
Comment (anila-intelops) [reply agent/kubviz/k8smetrics_agent.go]: changed from wg.Add(8) to wg.Add(9)

File: agent/kubviz/trivy_sbom.go (hunk 3 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: removed comment lines

File: agent/kubviz/trivy_sbom.go (hunk 4 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: Printf added instead of Fatalf

File: agent/kubviz/trivy_sbom.go (hunk 5 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: Printf added instead of Fatalf

File: agent/kubviz/trivy_sbom.go (hunk 6 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: stars removed

File: agent/kubviz/trivy_sbom.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+			command := fmt.Sprintf("trivy image --format cyclonedx %s", image.PullableImage)
+			out, err := executeCommandSbom(ctx, command)
+
+			if ctx.Err() == context.DeadlineExceeded {
+				log.Printf("*************Command execution timeout for image %s", image.PullableImage)
+				return // Move on to the next image
+			}
+
+			if err != nil {
+				log.Printf("*************Error executing Trivy for image %s: %v", image.PullableImage, err)
```
[the remark this answers is not in the document]
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: stars removed

File: agent/kubviz/trivy_sbom.go (hunk 7 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: stars removed

File: agent/kubviz/trivy_sbom.go (hunk 8 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: removed comment line

File: agent/kubviz/trivy_sbom.go (hunk 9 above)
Comment (anila-intelops) [reply agent/kubviz/trivy_sbom.go]: removed comment line

File: client/pkg/clickhouse/statements.go (hunk 10 above)
Comment (anila-intelops) [reply client/pkg/clickhouse/statements.go]: added components and metadata

File: client/pkg/clickhouse/db_client.go [hunk 12]
```diff
@@ -209,6 +210,83 @@ func (c *DBClient) InsertDeletedAPI(deletedAPI model.DeletedAPI) {
...
+// 		result.Vulnerabilities,
+// 	); err != nil {
+// 		log.Fatal(err)
+// 	}
+// 	if err := tx.Commit(); err != nil {
+// 		log.Fatal(err)
+// 	}
+// 	stmt.Close()
+
+// }
```
Comment (avinashkna4) [inline client/pkg/clickhouse/db_client.go:238]: remove commented code

File: client/pkg/clickhouse/db_client.go [hunk 13]
```diff
@@ -209,6 +210,83 @@ func (c *DBClient) InsertDeletedAPI(deletedAPI model.DeletedAPI) {
...
+func (c *DBClient) InsertTrivySbomMetrics(metrics model.Reports) {
+	result := metrics.Report
+	for _, metaTool := range result.Metadata.Tools {
+		for _, com := range result.Components {
+			for _, comPro := range com.Properties {
+				for _, comHash := range com.Hashes {
+					for _, comLicense := range com.Licenses {
+						for _, depend := range result.Dependencies {
```
Comment (avinashkna4) [inline client/pkg/clickhouse/db_client.go:246]: why so many inner loops. Can you send me one sample json output and DB table content for the same ?

File: client/pkg/clickhouse/db_client.go (hunk 12 above)
Comment (an1l4) [reply client/pkg/clickhouse/db_client.go:238]: removed comment

File: client/pkg/clickhouse/db_client.go (hunk 13 above)
Comment (an1l4) [reply client/pkg/clickhouse/db_client.go:246]: reduced number of for loop

**Pull request closed without merging**